### PR TITLE
Address race conditions in source database

### PR DESCRIPTION
**Pull request by jmcphers**
The underlying issue addressed by this change is that the persistent (`per`) directory of the source database can be read and written to, without guards, by multiple RStudio sessions. When running in multiple session mode this isn't a problem since each session gets its own persistent directory.

https://github.com/rstudio/rstudio/blob/3dff46a66359df96a01e247a4a8a99c26cc1a5e2/src/cpp/session/SessionSourceDatabaseSupervisor.cpp#L71-L82

In desktop mode however there exists only a single persistent directory, even though multiple sessions can exist. This creates the potential for timing-related errors.

This change does two things to mitigate the problem:

**Use a stable folder for the source database**. Prior to this change, the source database got a unique ID on every session start. This meant that on restart or suspend/resume, the whole source database was pushed into the persistent directory on the way down, and rehydrated from the persistent directory on the way up, into a fresh directory with a new ID. 

After the change, the name of the source database folder is stable through suspend/resume. The folder therefore just stays in place until the session is actually terminated (quit). Since sessions no longer read or write from the persistent folder during the ordinary course of operations, it's not necessary to guard access to that folder with concurrency primitives.

In order to distinguish source database folders which represent suspended (or restarting) sessions from those which represent crashed sessions, a `suspend_file` is written to the folder prior to suspending and removed when resuming. Folders containing these files aren't treated as crashed sessions when looking for an orphan session to adopt.

**Add a launcher token to the session.** There can be multiple desktop sessions running at once, so when a session is resumed, it needs to know which source database to reattach to. This information is supplied from the session launcher by generating a random "launcher token" and supplying it to the session when it (re)starts. 

The launcher token functions more or less identically to the session ID in multisession mode. It represents an instance of a session launcher (for instance, an instance of rstudio.exe). In fact, the analog is close enough that it can be used as a drop-in replacement for the active session ID in most places, and replaces the "virtual session ID" used previously. 

For consistency, we also use a launcher token in server mode. An entire server can use a single launcher token; in single session mode, the context ID already scopes the source database, and in multisession mode, the session ID replaces the launcher token as the identifier for the database.

Some caveats:

1. No attempt is made to ensure that launcher tokens are unique. The odds of this being a problem are vanishingly small as the set of active launchers is bounded by the number of open RStudio windows (and therefore the odds of a collision are infinitesimal), but we could guard against this if we think it's worth the additional complexity.

2. If for some reason a session suspends but is unable to resume, that session's source database will be permanently orphaned. Again, this seems very unlikely, but if we wanted to guard against it we could so by using a lock file owned by the session launcher.

3. As there are no concurrency guards on the persistent directory, it is still possible for there to be write conflicts here in pathological conditions, but as writes only happen on session quit, the worst case is that the next new session launch doesn't have any open documents (significantly more benign than having a current session lose its documents!). Again, we could guard against this with a lockfile if we felt it necessary.

4. Now that source database folders are stable, we could probably dispense with the per-session persistent folders used in multisession mode with some small amount of additional effort.

### CODE REVIEW

File: src/cpp/session/SessionSourceDatabaseSupervisor.cpp [hunk 1]
```diff
@@ -354,18 +351,29 @@ bool reclaimOrphanedSession(FilePath* pSessionDir)
 
    BOOST_FOREACH(const FilePath& sessionDir, sessionDirs)
    {
+      // if the suspend file exists, this session is only sleeping, not dead
+      if (sessionSuspendFilePath(sessionDir).exists())
```
Comment (kevinushey) [inline src/cpp/session/SessionSourceDatabaseSupervisor.cpp]: Would it be worth the additional complexity to assume that 'stale' files here are considered as orphans? E.g. if the file modify time is more than a minute old?

Comment (jmcphers) [inline src/cpp/session/SessionSourceDatabaseSupervisor.cpp]: These files have effectively unlimited age as sessions can be suspended indefinitely in server mode.

Comment (kevinushey) [inline src/cpp/session/SessionSourceDatabaseSupervisor.cpp]: Ah, right -- sorry, I was thinking specifically in the context of a restart (which is an explicit suspend + resume action); ie, thinking specifically if we can / should attempt to detect a failed restart.

**Comment by jjallaire** [comment]
LGTM but obviously this does carry some unknown risk of regressing other things. If you think that this isn't a huge liability (happens rarely) you could consider taking it a release with a longer burn-down / more testing (post 1.1). Let's discuss realtime later today.

File: src/cpp/session/SessionSourceDatabaseSupervisor.cpp (hunk 1 above)
Comment (jmcphers) [inline src/cpp/session/SessionSourceDatabaseSupervisor.cpp]: It adds to the complexity of the change, but yes, with some plumbing it's possible to distinguish a suspend for restart from an ordinary timeout-based (or server-initiated) suspend. Done here: https://github.com/rstudio/rstudio/commit/c5bd4974bcaac6d65dca05a455580800b36f9d0b

**Comment by jmcphers** [comment]
@jjallaire I'd prefer taking it for v1.1 for two reasons:

a) this feels like the right time to take higher risk bug fixes for v1.1 (as they'll have practically the entire release cycle to bake)

b) there are a small number of customers who have reported this; it'd be great to get a fix in their hands to see if we've actually gotten at the root of the problem or only fixed one particular case.

**Comment by jjallaire** [comment]
Okay, agreed. Let's take it for v1.1.

**Pull request merged**